clippy_lints/src/methods/waker_clone_wake.rs

```diff
@@ -0,0 +1,34 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::{match_def_path, paths};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::WAKER_CLONE_WAKE;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>, recv: &'tcx Expr<'_>) {
+    let ty = cx.typeck_results().expr_ty(recv);
+
+    if let Some(did) = ty.ty_adt_def()
+        && match_def_path(cx, did.did(), &paths::WAKER)
+        && let ExprKind::MethodCall(func, waker_ref, &[], _) = recv.kind
+        && func.ident.name == sym::clone
```

Comment:
We should probably use [`is_trait_method(cx, recv, sym::Clone)`](https://doc.rust-lang.org/stable/nightly-rustc/clippy_utils/fn.is_trait_method.html) here, instead of only checking the identifier.

It avoids linting in one edge-case where it shouldn't:
```rs
#![feature(noop_waker)]
use std::task::Waker;

struct MyWaker;
impl MyWaker {
  pub fn clone(&self) -> Waker {
    Waker::noop()
  }
}

MyWaker.clone().wake();
```
This would also be a good test.

(I don't think anyone is going to write code like this, but `MachineApplicable` suggestions are expected to always compile)